tracer/src/Datadog.Trace/Tagging/TagsList.cs

```diff
@@ -3,389 +3,134 @@
+                        {
+                            tags[i] = new KeyValuePair<string, string>(key, value);
+                        }
 
-                    // don't cast to IEnumerable so we can use the struct enumerator from List<T>
-                    foreach (var tag in tags)
-                    {
-                        WriteTag(ref bytes, ref offset, tag.Key, tag.Value, tagProcessors);
+                        return;
                     }
                 }
-            }
-
-            return count;
-        }
-
-        // Metrics
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        protected void WriteMetric(ref byte[] bytes, ref int offset, string key, double value, ITagProcessor[] tagProcessors)
-        {
-            if (tagProcessors is not null)
-            {
-                for (var i = 0; i < tagProcessors.Length; i++)
-                {
-                    tagProcessors[i]?.ProcessMetric(ref key, ref value);
-                }
-            }
-
-            offset += MessagePackBinary.WriteString(ref bytes, offset, key);
-            offset += MessagePackBinary.WriteDouble(ref bytes, offset, value);
-        }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        protected void WriteMetric(ref byte[] bytes, ref int offset, byte[] keyBytes, double value, ITagProcessor[] tagProcessors)
-        {
-            if (tagProcessors is not null)
-            {
-                string key = null;
-                for (var i = 0; i < tagProcessors.Length; i++)
+                // If we get there, the tag wasn't in the collection
+                if (value != null)
                 {
-                    tagProcessors[i]?.ProcessMetric(ref key, ref value);
+                    tags.Add(new KeyValuePair<string, string>(key, value));
                 }
             }
-
-            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, keyBytes);
-            offset += MessagePackBinary.WriteDouble(ref bytes, offset, value);
         }
 
-        private int WriteMetrics(ref byte[] bytes, int offset, Span span, ITagProcessor[] tagProcessors)
+        public virtual void EnumerateTags<TProcessor>(ref TProcessor processor)
+            where TProcessor : struct, IItemProcessor<string>
         {
-            int originalOffset = offset;
-
-            // Start of "metrics" dictionary. Do not add any numeric tags before this line.
-            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, MetricsBytes);
-
-            int count = 0;
-
-            // We don't know the final count yet, write a fixed-size header and note the offset
-            var countOffset = offset;
-            offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0);
-
-            // write "custom" span-level numeric tags (from list of KVPs)
-            var metrics = Metrics;
-            if (metrics != null)
+            var values = Volatile.Read(ref _tags);
```

Comment:
nit - I would not name `values` a structure representing a list of kvp. You could maybe reuse `tags` as in other functions

Reply:
good catch, done in https://github.com/DataDog/dd-trace-dotnet/pull/2753/commits/7a50cec75953586effb8c3007516ceb13e8ce4db